File: contest/snippets/haskell/math/powMod.hs
powMod :: Int -> Int -> Int -> Int
powMod a b p
    | b == 0 = 1
    | odd b  = ( a * (powMod a (b - 1) p)) `mod` p
    | even b = ( d * d ) `mod` p where
        d = powMod a (b `div` 2) p

File: contest/snippets/python/math/pow_mod.py
def pow_mod(a, k, M):
    if k == 0:
        return 1

    t = pow_mod(a, k // 2, M)
    res = (t * t) % M
    if k % 2 == 1:
        res = (res * a) % M

    return res
